### PR TITLE
[preferences] add the ability to 'save and retry' when prompted for dirty preferences

### PR DESCRIPTION
**Pull request by vince-fugnitto**
Added a small functionality to the message displayed to users when there are dirty preference editors and they attempt to make an edit through the preferences-tree. Users will have the ability to select the `Save and Retry` button to successfully save their preference editors and re-attempt to apply a preference edit.

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by marechal-p** [comment]
It looks bugged, I have to trigger the `Save and Retry` twice.

Also I think that when setting preferences from the tree should not directly write on disk, but rather append to what is already in the editor, user can then save or not.

**Comment by vince-fugnitto** [comment]
@marechal-p 

> It looks bugged, I have to trigger the `Save and Retry` twice.

I'll take a look and re-test :)

> Also I think that setting preferences from the tree should not directly write on disk, but rather append to what is already in the editor, user can then save or not.

This is the same behavior in vscode, if a preference editor is dirty, the command `save and retry` simply saves the editor, and retries the edit. Do you think we should not do the same?

**Comment by marechal-p** [comment]
Turns out it was the update taking too long, clicking once works.

> This is the same behavior in vscode, if a preference editor is dirty, the command `save and retry` simply saves the editor, and retries the edit. Do you think we should not do the same?

This is just my opinion, but our widgets are different by nature. In ours, my opinion is that the way to input the preferences is through the json file directly, and using the tree is just another way of contributing to that file. Hence why I see using the tree actually writing in the editor's buffer rather than trying to access the disk and prompting something when the content of the file you are editing is right in front of you. But that's just my way of seeing it.

Overall the change seems to do what it is meant to do, I just don't like it.

**Comment by vince-fugnitto** [comment]
> Overall the change seems to do what it is meant to do, I just don't like it.

Maybe we should update the logic in the `preferences-tree` to simply update the json (append) without actually saving it, meaning the updates will only be in the buffer like you said and not saved to disk.

Also, the `Save and Retry` is a user interaction. It's up to the user to use it and I don't attempt to save sneakily. I assume they'd understand that a save and retry means save all dirty changes and retry the edit.

File: packages/preferences/src/browser/preferences-tree-widget.ts [hunk 1]
```diff
@@ -140,7 +140,12 @@ export class PreferencesContainer extends SplitPanel implements ApplicationShell
             const preferenceName = Object.keys(value)[0];
             const preferenceValue = value[preferenceName];
             if (this.dirty) {
-                this.messageService.warn('Preferences editor(s) has/have unsaved changes');
+                this.showDirtyMessage().then(async selected => {
```
Comment (marechal-p) [inline packages/preferences/src/browser/preferences-tree-widget.ts]: There is no async code.

File: packages/preferences/src/browser/preferences-tree-widget.ts [hunk 2]
```diff
@@ -245,6 +250,17 @@ export class PreferencesContainer extends SplitPanel implements ApplicationShell
...
+
+    /**
+     * Show a message when there are unsaved preference editors.
+     * Users will have the opportunity to save and retry an edit.
+     * @returns a Promise that resolves to the user's selection or `undefined`.
+     */
+    private showDirtyMessage(): Promise<string | undefined> {
+        const MESSAGE: string = 'Preferences editor(s) has/have unsaved changes';
+        const SAVE_RETRY: string = 'Save and Retry';
+        return Promise.resolve(this.messageService.warn(MESSAGE, SAVE_RETRY));
```
Comment (marechal-p) [inline packages/preferences/src/browser/preferences-tree-widget.ts]: Do we need the variables for the text? It is not even reused elsewhere, we should be able to do without.

Also, `this.messageService.warn()` already returns a promise, please remove `Promise.resolve`.

**Comment by lmcbout** [comment]
It looks like the "auto.save" does not applu to the user preferences unless it is set in the "User" tab preferences. If you set it at the "Workspace" level, the "User" preference are not saved automatically

**Comment by vince-fugnitto** [comment]
@marechal-p can you re-try, I think I addressed the buggy behavior :)

**Comment by vince-fugnitto** [comment]
@elaihau if you get the chance could you test as well? I don't have objections in closing the PR if its not necessary :)

**Comment by vince-fugnitto** [comment]
@elaihau I'll close the PR for the moment as I don't believe it adds much functionality and/or is a crucial feature. If ever I have the bandwidth I'll revisit it more closely :)

**Pull request closed without merging**